src/core/Akka.Remote/Endpoint.cs

```diff
@@ -411,147 +414,174 @@ protected override void PostRestart(Exception reason)
-                    UidConfirmed = true;
-                    if (Uid.HasValue && Uid.Value != g.Uid) Reset();
-                    else UnstashAcks();
-                    Uid = _refuseUid;
-                })
-                .With<EndpointWriter.StopReading>(stopped =>
-                {
-                    _writer.Forward(stopped); //forward the request
-                });
+        private void GoToIdle()
+        {
+            if (_bufferWasInUse && _maxSilenceTimer == null)
+                _maxSilenceTimer =
+                    Context.System.Scheduler.ScheduleTellOnceCancelable(_settings.QuarantineSilentSystemTimeout, Self,
+                        TooLongIdle.Instance, Self);
+            Become(IdleBehavior);
         }
 
-        protected void Gated(object message)
+        private void GoToActive()
         {
-            message.Match()
-                .With<Terminated>(
-                    terminated => Context.System.Scheduler.ScheduleTellOnce(_settings.RetryGateClosedFor, Self, new Ungate(), Self))
-                .With<Ungate>(ungate =>
-                {
-                    if (_resendBuffer.NonAcked.Count > 0 || _resendBuffer.Nacked.Count > 0)
-                    {
-                        // If we talk to a system we have not talked to before (or has given up talking to in the past) stop
-                        // system delivery attempts after the specified time. This act will drop the pending system messages and gate the
-                        // remote address at the EndpointManager level stopping this actor. In case the remote system becomes reachable
-                        // again it will be immediately quarantined due to out-of-sync system message buffer and becomes quarantined.
-                        // In other words, this action is safe.
-                        if (!UidConfirmed && BailoutAt.IsOverdue)
-                        {
-                            throw new InvalidAddressAssociationException(_localAddress, _remoteAddress,
-                                new TimeoutException("Delivery of system messages timed out and they were dropped"));
-                        }
+            _maxSilenceTimer?.Cancel();
+            _maxSilenceTimer = null;
+            Become(Receiving);
+        }
 
-                        _writer = CreateWriter();
-                        //Resending will be triggered by the incoming GotUid message after the connection finished
-                        Context.Become(OnReceive);
-                    }
-                    else
-                    {
-                        Context.Become(Idle);
-                    }
-                })
-                .With<EndpointManager.Send>(send =>
+        protected void Gated(bool writerTerminated, bool earlyUngateRequested)
+        {
+            Receive<Terminated>(terminated =>
+            {
+                if (!writerTerminated)
```

Comment:
If we receive a `Terminated` message in this state and `writerTerminated` == `false`, then this message could have only come from the `EndpointWriter` terminating. Going to change behavior to `Gated(true, earlyUngateRequested)` then.

If `earlyUngateRequested` == `true`, then we will send ourselves an `Ungate` command which will re-open the association.

Otherwise, we schedule a `Ungate` message to be delivered to ourselves in the future per the normal `RemoteSettings.RetryGateClosedFor` duration.